keps/sig-apps/0008-sidecarcontainer.md

```diff
@@ -0,0 +1,145 @@
+
+Here are some examples of the main problems:
+
+### Jobs
+ If you have a Job with two containers one of which is actually doing the main processing of the job and the other is just facilitating it, you encounter a problem when the main process finishes; your sidecar container will carry on running so the job will never finish.
+
+The only way around this problem is to manage the sidecar container's lifecycle manually and arrange for it to exit when the main container exits. This is typically achieved by building an ad-hoc signalling mechanism to communicate completion status between containers. Common implementations use a shared scratch volume mounted into all pods, where lifecycle status can be communicated by creating and watching for the presence of files. This pattern has several disadvantages:
+
+* Repetitive lifecycle logic must be rewritten in each instance a sidecar is deployed.
+* Third-party containers typically require a wrapper to add this behaviour, normally provided via an entrypoint wrapper script implemented in the k8s container spec. This adds undesirable overhead and introduces repetition between the k8s and upstream container image specs.
+* The wrapping typically requires the presence of a shell in the container image, so this pattern does not work for minimal containers which ship without a toolchain.
+
+### Startup
+An application that has a proxy container acting as a sidecar may fail when it starts up as it's unable to communicate until its proxy has started up successfully. Readiness probes don't help if the application is trying to talk outbound.
+
+### Shutdown
+Applications that rely on sidecars may experience a high amount of errors when shutting down as the sidecar may terminate before the application has finished what it's doing.
+
+
+## Goals
+
+Solve issues so that they don't require application modification:
+[25908](https://github.com/kubernetes/kubernetes/issues/25908) - Job completion
+[65502](https://github.com/kubernetes/kubernetes/issues/65502) - Container startup dependencies
+
+## Non-Goals
+
+Allowing multiple containers to run at once during the init phase.
+
+## Proposal
+
+Create a way to define containers as sidecars, this will be an additional field to the Container Spec: `sidecar: true`.
+e.g:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: myapp-pod
+  labels:
+    app: myapp
+spec:
+  containers:
+  - name: myapp
+    image: myapp
+    command: ['do something']
+  - name: sidecar
+    image: sidecar-image
+    sidecar: true
+    command: ["do something to help my app"]
+
+```
+Sidecars will be started before normal containers but after init, so that they are ready before your main processes start.
+
+This will change the Pod startup to look like this:
+* Init containers start
+* Init containers finish
+* Sidecars start
+* Containers start
+
+During pod termination sidecars will be terminated last:
+* Containers sent SIGTERM
+* Sidecars sent SIGTERM
```

Comment:
Does this imply that user-containers must terminate before sidecars are sent SIGTERM at all?  Or that some time (how much?) is waited between user-containers and sidecars?  Or just that the ordering is strict (which seems relatively pointless to me) ?

Reply:
@thockin so sidecars would get sent a SIGTERM after the containers have exited, this is to ensure that your applications still have the sidecars it relies on while shutting down. Currently all containers will get SIGKILL if termination grace period is elapsed. I don't know if there's any value in separating the grace period for sidecars and containers.

Reply:
What you say makes sense, the doc is not clear.

Reply:
@thockin ok I shall update it to make it clearer